proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8244,20 +8258,21 @@ lemma refill_update_valid_refills:
   apply presburger
   done
 
-lemma schedule_used_sum[simp]:
-  "refills_sum (schedule_used b refills new) = refills_sum (refills @ [new])"
-  apply (cases refills)
-   apply (clarsimp simp: schedule_used_def)
+lemma schedule_used_sum:
+  "\<lbrace>\<lambda>s. pred_map (\<lambda>cfg. refills_sum ((scrc_refills cfg) @ [new]) = t) (sc_refill_cfgs_of s) (cur_sc s)
+        \<and> sc_ptr = cur_sc s\<rbrace>
+   schedule_used sc_ptr new
+   \<lbrace>\<lambda>_ s. pred_map (\<lambda>cfg. refills_sum (scrc_refills cfg) = t) (sc_refill_cfgs_of s) (cur_sc s)\<rbrace>"
```

Comment:
It's very hard to say without actually playing around with it, but would it be easier if these lemmas about `schedule_used` were generalised away from `cur_sc`?

Reply:
It certainly wouldn't be any easier if these weren't `cur_sc`. I found it very annoying to get this all right, and I dreaded trying to use more general versions along with `wps` (which I've never had a good handle on). If you'd like me to take another look, I can, or maybe I can make a JIRA issue for it